pkg/snapshot/cloudsnap.go

```diff
@@ -0,0 +1,140 @@
+package snapshot
+
+import (
+	"fmt"
+
+	"github.com/libopenstorage/openstorage/api"
+	"github.com/libopenstorage/openstorage/volume"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	configCred = "CredId"
+)
+
+type cloudSnapshotPlugin struct {
+	Plugin
+	log    logrus.FieldLogger
+	credID string
+}
+
+func (c *cloudSnapshotPlugin) Init(config map[string]string) error {
+	c.credID = config[configCred]
+
+	c.log.Infof("Init'ing portworx cloud snapshot with credID %v", c.credID)
+	return nil
+}
+
+func (c *cloudSnapshotPlugin) CreateVolumeFromSnapshot(snapshotID, volumeType, volumeAZ string, iops *int64) (string, error) {
+	volDriver, err := getVolumeDriver()
+	if err != nil {
+		return "", err
+	}
+
+	// Enumerating can be expensive but we need to do it to get the original
+	// volume name. Ark already has it so it can pass it down to us.
+	// CloudBackupRestore can also be updated to restore to the original volume
+	// name.
+	enumRequest := &api.CloudBackupEnumerateRequest{}
+	enumRequest.CredentialUUID = c.credID
+	enumResponse, err := volDriver.CloudBackupEnumerate(enumRequest)
+	if err != nil {
+		return "", err
+	}
+
+	volumeName := ""
+	for _, backup := range enumResponse.Backups {
+		if backup.ID == snapshotID {
+			volumeName = backup.SrcVolumeName
+			break
+		}
+	}
+
+	if volumeName == "" {
+		return "", fmt.Errorf("Couldn't find volume name from cloudsnap")
+	}
+
+	response, err := volDriver.CloudBackupRestore(&api.CloudBackupRestoreRequest{
+		ID:                snapshotID,
+		CredentialUUID:    c.credID,
+		RestoreVolumeName: volumeName,
```

Comment:
is this restore in a different cluster than where the backup was created? If not, this will most likely fail due to volumeName collision.

Reply:
Restore can be done to the same cluster or different cluster. The restore is expected to fail if the volume with the same name already exists.